app/gui2/src/util/containers.ts

```diff
@@ -0,0 +1,246 @@
+
+  /** Set the value for the given key. */
+  set(key: Key, value: Value): this {
+    const innerKey = this.keyMapper(key)
+    this.map.set(innerKey, [key, value])
+    return this
+  }
+
+  /** Get the value for the given key, or `undefined` if it does not exist. */
+  get(key: Key): Value | undefined {
+    const innerKey = this.keyMapper(key)
+    const entry = this.map.get(innerKey)
+    return entry ? entry[1] : undefined
+  }
+
+  /** Check if the map contains a value for the given key. */
+  has(key: Key): boolean {
+    const innerKey = this.keyMapper(key)
+    return this.map.has(innerKey)
+  }
+
+  /** Remove the value for the given key. */
+  delete(key: Key): boolean {
+    const innerKey = this.keyMapper(key)
+    return this.map.delete(innerKey)
+  }
+
+  /** Remove all values from the map. */
+  clear(): void {
+    this.map.clear()
+  }
+
+  /** Get the number of values in the map. */
+  get size(): number {
+    return this.map.size
+  }
+
+  /** Iterate over the values in the map. */
+  *[Symbol.iterator](): IterableIterator<[Key, Value]> {
+    for (const [_innerKey, [key, value]] of this.map.entries()) {
+      yield [key, value]
+    }
+  }
+}
+
+/**
+ * Set that supports Object-based keys.
+ *
+ * Internally keys are converted to strings using the provided {@link keyMapper} function and then compared.
+ *
+ * @template Key The type of the keys.
+ */
+export class ObjectKeyedSet<Key extends Object> {
+  /** The inner set that stores the keys. */
+  private readonly set = new Set<string>()
+
+  /** Construct a new set, optionally setting a custom key mapper.
+   *
+   * @param keyMapper The function that maps the keys to strings. By default, strings are generated using {@link JSON.stringify}.
+   *
+   */
+  constructor(private readonly keyMapper: (key: Key) => string = JSON.stringify) {}
+
+  /** Add the given key to the set. */
+  add(key: Key): this {
+    const innerKey = this.keyMapper(key)
+    this.set.add(innerKey)
+    return this
+  }
+
+  /** Check if the set contains the given key. */
+  has(key: Key): boolean {
+    const innerKey = this.keyMapper(key)
+    return this.set.has(innerKey)
+  }
+
+  /** Remove the given key from the set. */
+  delete(key: Key): boolean {
+    const innerKey = this.keyMapper(key)
+    return this.set.delete(innerKey)
+  }
+
+  /** Remove all keys from the set. */
+  clear(): void {
+    this.set.clear()
+  }
+
+  /** Get the number of keys in the set. */
+  get size(): number {
+    return this.set.size
+  }
+
+  /** Iterate over the keys in the set. */
+  *[Symbol.iterator](): IterableIterator<Key> {
+    for (const innerKey of this.set) {
+      yield JSON.parse(innerKey) as Key
+    }
+  }
+}
+
+/** Stack that always has at least one element.
+ *
+ * It is meant to be used with scope-based operations, thus it does not provide direct `push` and `pop` methods.
+ */
+export class NonEmptyStack<T> {
+  /** The "actual" stack of elements. */
+  private readonly stack: T[]
```

Comment:
Non-empty array can be represented with type `[T, ...T[]]`. We already have a type alias for that in `util/array.ts`.

Reply:
Used it.